app/src/main/java/com/kickstarter/viewmodels/MessagesViewModel.java

```diff
@@ -0,0 +1,155 @@
+import com.kickstarter.models.Backing;
+import com.kickstarter.models.Message;
+import com.kickstarter.models.MessageThread;
+import com.kickstarter.models.Project;
+import com.kickstarter.services.ApiClientType;
+import com.kickstarter.services.apiresponses.MessageThreadEnvelope;
+import com.kickstarter.ui.IntentKey;
+import com.kickstarter.ui.activities.MessagesActivity;
+
+import java.util.List;
+
+import rx.Notification;
+import rx.Observable;
+import rx.subjects.BehaviorSubject;
+import rx.subjects.PublishSubject;
+
+import static com.kickstarter.libs.rx.transformers.Transformers.values;
+
+public interface MessagesViewModel {
+
+  interface Inputs {
+    /** Call when the message edit text changes. */
+    void messageEditTextChanged(String messageBody);
+  }
+
+  interface Outputs {
+    /** Emits the backing and project to populate the backing info header. */
+    Observable<Pair<Backing, Project>> backingAndProject();
+
+    /** Emits a boolean that determines if the backing info view should be hidden. */
+    Observable<Boolean> backingInfoViewHidden();
+
+    /** Emits the creator name to be displayed. */
+    Observable<String> creatorNameTextViewText();
+
+    /** Emits a list of messages to be displayed. */
+    Observable<List<Message>> messages();
+
+    /** Emits the project name to be displayed. */
+    Observable<String> projectNameTextViewText();
+  }
+
+  final class ViewModel extends ActivityViewModel<MessagesActivity> implements Inputs, Outputs {
+    private final ApiClientType client;
+    private final CurrentUserType currentUser;
+
+    public ViewModel(final @NonNull Environment environment) {
+      super(environment);
+
+      this.client = environment.apiClient();
+      this.currentUser = environment.currentUser();
+
+      final Observable<MessageThread> messageThread = intent()
+        .take(1)
+        .map(i -> i.getParcelableExtra(IntentKey.MESSAGE_THREAD))
+        .ofType(MessageThread.class);
+
+      final Observable<Notification<MessageThreadEnvelope>> envelopeNotification = messageThread
+        .switchMap(thread -> this.client.fetchMessagesForThread(thread).materialize())
+        .share();
+
+      final Observable<MessageThreadEnvelope> messageThreadEnvelope = envelopeNotification.compose(values());
+
+      final Observable<Notification<Backing>> backingNotification = Observable.combineLatest(
+        messageThread,
+        this.currentUser.observable(),
+        Pair::create
+      )
+        .filter(mu -> mu.first.backing() == null)
+        .switchMap(mu ->
+          mu.first.project().isBacking()
+            ? this.client.fetchProjectBacking(mu.first.project(), mu.second).materialize()
+            : this.client.fetchProjectBacking(mu.first.project(), mu.first.participant()).materialize()
+        )
+        .share();
+
+      final Observable<Backing> backing = Observable.merge(
+        messageThread.map(MessageThread::backing).filter(ObjectUtils::isNotNull),
```

Comment:
looks like we could drop this `filter` if yer doing the other `filter` down below